src/charm.py

```diff
@@ -0,0 +1,1017 @@
+    WaitingStatus,
+)
+from ops.pebble import ChangeError, Error, ExecError, Layer
+
+from hydra_cli import HydraCLI
+from utils import normalise_url, remove_none_values
+
+logger = logging.getLogger(__name__)
+
+EXTRA_USER_ROLES = "SUPERUSER"
+HYDRA_ADMIN_PORT = 4445
+HYDRA_PUBLIC_PORT = 4444
+SUPPORTED_SCOPES = ["openid", "profile", "email", "phone"]
+PEER = "hydra"
+LOG_LEVELS = ["panic", "fatal", "error", "warn", "info", "debug", "trace"]
+DB_MIGRATION_VERSION_KEY = "migration_version"
+COOKIE_SECRET_KEY = "cookie"
+COOKIE_SECRET_LABEL = "cookiesecret"
+SYSTEM_SECRET_KEY = "system"
+SYSTEM_SECRET_LABEL = "systemsecret"
+
+
+class HydraCharm(CharmBase):
+    """Charmed Ory Hydra."""
+
+    def __init__(self, *args: Any) -> None:
+        super().__init__(*args)
+
+        self._container_name = "hydra"
+        self._container = self.unit.get_container(self._container_name)
+        self._hydra_config_path = "/etc/config/hydra.yaml"
+        self._db_name = f"{self.model.name}_{self.app.name}"
+        self._db_relation_name = "pg-database"
+        self._login_ui_relation_name = "ui-endpoint-info"
+        self._prometheus_scrape_relation_name = "metrics-endpoint"
+        self._loki_push_api_relation_name = "logging"
+        self._grafana_dashboard_relation_name = "grafana-dashboard"
+        self._tracing_relation_name = "tracing"
+        self._hydra_service_command = "hydra serve all"
+        self._log_dir = Path("/var/log")
+        self._log_path = self._log_dir / "hydra.log"
+
+        self._hydra_cli = HydraCLI(
+            f"http://localhost:{HYDRA_ADMIN_PORT}", self._container, self._hydra_config_path
+        )
+
+        self.service_patcher = KubernetesServicePatch(
+            self, [("hydra-admin", HYDRA_ADMIN_PORT), ("hydra-public", HYDRA_PUBLIC_PORT)]
+        )
+
+        self.database = DatabaseRequires(
+            self,
+            relation_name=self._db_relation_name,
+            database_name=self._db_name,
+            extra_user_roles=EXTRA_USER_ROLES,
+        )
+        self.admin_ingress = IngressPerAppRequirer(
+            self,
+            relation_name="admin-ingress",
+            port=HYDRA_ADMIN_PORT,
+            strip_prefix=True,
+            redirect_https=False,
+        )
+        self.public_ingress = IngressPerAppRequirer(
+            self,
+            relation_name="public-ingress",
+            port=HYDRA_PUBLIC_PORT,
+            strip_prefix=True,
+            redirect_https=False,
+        )
+        self.oauth = OAuthProvider(self)
+
+        self.login_ui_endpoints = LoginUIEndpointsRequirer(
+            self, relation_name=self._login_ui_relation_name
+        )
+
+        self.endpoints_provider = HydraEndpointsProvider(self)
+
+        self.metrics_endpoint = MetricsEndpointProvider(
+            self,
+            relation_name=self._prometheus_scrape_relation_name,
+            jobs=[
+                {
+                    "metrics_path": "/admin/metrics/prometheus",
```

Comment:
```suggestion
                {
                    "job_name": "hydra_metrics",
                    "metrics_path": "/admin/metrics/prometheus",
```

Maybe add a job name here. As if not provided, the Prometheus charm will automatically generate one and the generated name may change with unit changes, disrupting the continuity of the metrics. see https://github.com/canonical/hydra-operator/blob/565596c0b84ea0e23f910859c8c9380cd6ef4765/lib/charms/prometheus_k8s/v0/prometheus_scrape.py#L167-L170